engine/src/main/java/io/zeebe/engine/nwe/container/ProcessProcessor.java

```diff
@@ -22,69 +26,173 @@
   public void onCompleted(
-      final ExecutableFlowElementContainer element, final BpmnElementContext context) {}
+      final ExecutableFlowElementContainer element, final BpmnElementContext context) {
+
+    final var parentWorkflowInstanceKey = context.getParentWorkflowInstanceKey();
+    final var parentElementInstanceKey = context.getParentElementInstanceKey();
+
+    if (parentWorkflowInstanceKey > 0) {
+      // workflow instance is created by a call activity
+
+      // TODO (saig0): move responsibility to call activity (#4473)
+      final var parentElementInstance = stateBehavior.getParentElementInstanceContext(context);
+
+      if (parentElementInstance != null
+          && parentElementInstance.getIntent() == WorkflowInstanceIntent.ELEMENT_ACTIVATED) {
+        // complete the corresponding call activity
+
+        stateTransitionBehavior.transitionToCompleting(parentElementInstance);
+
+        // propagate the variables to the parent
+        final var variablesState = stateBehavior.getVariablesState();
+
+        final var variables =
+            variablesState.getVariablesAsDocument(context.getElementInstanceKey());
+        variablesState.setTemporaryVariables(parentElementInstanceKey, variables);
+      }
+    }
+
+    if (element.hasNoneStartEvent()) {
+      workflowResultSenderBehavior.sendResult(context);
+    }
+
+    if (element.hasMessageStartEvent()) {
+      bufferedMessageStartEventBehavior.correlateMessage(context);
+    }
+
+    stateBehavior.removeElementInstance(context);
+  }
 
   @Override
   public void onTerminating(
-      final ExecutableFlowElementContainer element, final BpmnElementContext context) {}
+      final ExecutableFlowElementContainer element, final BpmnElementContext context) {
+
+    eventSubscriptionBehavior.unsubscribeFromEvents(context);
+
+    stateTransitionBehavior.terminateChildInstances(context);
+  }
 
   @Override
   public void onTerminated(
-      final ExecutableFlowElementContainer element, final BpmnElementContext context) {}
+      final ExecutableFlowElementContainer element, final BpmnElementContext context) {
+
+    incidentBehavior.resolveIncidents(context);
+
+    final var parentWorkflowInstanceKey = context.getParentWorkflowInstanceKey();
+
+    if (parentWorkflowInstanceKey > 0) {
+      // workflow instance is created by a call activity
+
+      // TODO (saig0): move responsibility to call activity (#4473)
+      final var parentElementInstance = stateBehavior.getParentElementInstanceContext(context);
+
+      if (parentElementInstance != null
+          && parentElementInstance.getIntent() == WorkflowInstanceIntent.ELEMENT_TERMINATING) {
+        // terminate the corresponding call activity
+
+        stateTransitionBehavior.transitionToTerminated(parentElementInstance);
+      }
+    }
+
+    if (element.hasMessageStartEvent()) {
+      bufferedMessageStartEventBehavior.correlateMessage(context);
+    }
+
+    stateBehavior.removeElementInstance(context);
```

Comment:
Does termination not have to be communicated as a workflow result?

Reply:
No. Currently, the result is only sent when the workflow instance is completed.